pkg/relay/ethereum/events.go

```diff
@@ -16,227 +16,182 @@ import (
+		height := clienttypes.NewHeight(0, log.BlockNumber)
+
+		var sendPacket ibchandler.IbchandlerSendPacket
+		if err := abiIBCHandler.UnpackIntoInterface(&sendPacket, "SendPacket", log.Data); err != nil {
 			return nil, err
-		} else {
-			if l := len(values); l != 6 {
-				return nil, fmt.Errorf("unexpected values length: expected=%v actual=%v", 6, l)
-			}
-			pSequence := values[0].(uint64)
-			pSourcePortID := values[1].(string)
-			pSourceChannel := values[2].(string)
-			pTimeoutHeight := values[3].(struct {
-				RevisionNumber uint64 "json:\"revision_number\""
-				RevisionHeight uint64 "json:\"revision_height\""
-			})
-			pTimeoutTimestamp := values[4].(uint64)
-			pData := values[5].([]uint8)
-
-			if pSequence == sequence && pSourcePortID == sourcePortID && pSourceChannel == sourceChannel {
-				return &channeltypes.Packet{
-					Sequence:           pSequence,
-					SourcePort:         pSourcePortID,
-					SourceChannel:      pSourceChannel,
-					DestinationPort:    channel.Counterparty.PortId,
-					DestinationChannel: channel.Counterparty.ChannelId,
-					Data:               pData,
-					TimeoutHeight:      clienttypes.Height{RevisionNumber: pTimeoutHeight.RevisionNumber, RevisionHeight: pTimeoutHeight.RevisionHeight},
-					TimeoutTimestamp:   pTimeoutTimestamp,
-				}, nil
-			}
 		}
+
+		packet := &core.PacketInfo{
+			Packet: channeltypes.NewPacket(
+				sendPacket.Data,
+				sendPacket.Sequence,
+				sendPacket.SourcePort,
+				sendPacket.SourceChannel,
+				dstPortID,
+				dstChannelID,
+				clienttypes.Height(sendPacket.TimeoutHeight),
+				sendPacket.TimeoutTimestamp,
+			),
+			EventHeight: height,
+		}
+		packets = append(packets, packet)
 	}
 
-	return nil, fmt.Errorf("packet not found: sourcePortID=%v sourceChannel=%v sequence=%v", sourcePortID, sourceChannel, sequence)
+	return packets, nil
 }
 
-// getAllPackets returns all packets from events
-func (chain *Chain) getAllPackets(
-	ctx core.QueryContext,
-	sourcePortID string,
-	sourceChannel string,
-) ([]*channeltypes.Packet, error) {
-	channel, found, err := chain.ibcHandler.GetChannel(
-		chain.callOptsFromQueryContext(ctx),
-		sourcePortID, sourceChannel,
-	)
+func (chain *Chain) findReceivedPackets(ctx core.QueryContext, fromHeight uint64) (core.PacketInfoList, error) {
+	recvPacketEvents, err := chain.findRecvPacketEvents(ctx, fromHeight)
 	if err != nil {
 		return nil, err
-	} else if !found {
-		return nil, fmt.Errorf("channel not found: sourcePortID=%v sourceChannel=%v", sourcePortID, sourceChannel)
 	}
 
-	query := ethereum.FilterQuery{
-		FromBlock: big.NewInt(0),
-		ToBlock:   new(big.Int).SetUint64(ctx.Height().GetRevisionHeight()),
-		Addresses: []common.Address{
-			chain.config.IBCAddress(),
-		},
-		Topics: [][]common.Hash{{
-			abiSendPacket.ID,
-		}},
-	}
-	logs, err := chain.client.FilterLogs(ctx.Context(), query)
+	writeAckEvents, err := chain.findWriteAckEvents(ctx, fromHeight)
```

Comment:
Considering the order of receive-packet and write-acknowledgment, can we give `recvPacketEvents[0].Raw.BlockNumber` instead of `fromHeight`?

Reply:
Right. In my opinion, it feels a bit like premature optimization, but OK I apply the fix for this.

Reply:
Fixed: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/6/commits/069ceafc93d35d1f57ec88c23193b7983367e2ae